File: ios/MySpeedPuzzling/Billing/BillingBridge.swift
import Foundation
import WebKit
import StoreKit

/// JavaScript bridge for in-app purchases
/// Receives purchase requests from web JavaScript and initiates native StoreKit flows
class BillingBridge: NSObject, WKScriptMessageHandler {
    weak var webView: WKWebView?

    @MainActor
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == "billing",
              let body = message.body as? [String: Any],
              let action = body["action"] as? String else {
            return
        }

        switch action {
        case "purchase":
            if let productId = body["productId"] as? String {
                Task {
                    await handlePurchase(productId: productId)
                }
            }
        case "restore":
            Task {
                await handleRestore()
            }
        case "checkSubscription":
            Task {
                await checkSubscriptionStatus()
            }
        default:
            break
        }
    }

    @MainActor
    private func handlePurchase(productId: String) async {
        let storeKit = StoreKitManager.shared

        // Load products if not already loaded
        if storeKit.products.isEmpty {
            await storeKit.loadProducts()
        }

        guard let product = storeKit.products.first(where: { $0.id == productId }) else {
            sendPurchaseError("Product not found: \(productId)")
            return
        }

        do {
            if let transaction = try await storeKit.purchase(product) {
                sendPurchaseSuccess(productId: productId, transactionId: String(transaction.id))
            } else {
                // User cancelled or pending
                sendPurchaseCancelled()
            }
        } catch {
            sendPurchaseError(error.localizedDescription)
        }
    }

    @MainActor
    private func handleRestore() async {
        let storeKit = StoreKitManager.shared
        await storeKit.restorePurchases()

        let hasSubscription = await storeKit.hasActiveSubscription()
        if hasSubscription {
            sendRestoreSuccess()
        } else {
            sendRestoreError("No active subscription found")
        }
    }

    @MainActor
    private func checkSubscriptionStatus() async {
        let hasSubscription = await StoreKitManager.shared.hasActiveSubscription()
        sendSubscriptionStatus(active: hasSubscription)
    }

    // MARK: - JavaScript Callbacks

    private func sendPurchaseSuccess(productId: String, transactionId: String) {
        let js = "window.onIosPurchaseSuccess && window.onIosPurchaseSuccess('\(productId)', '\(transactionId)')"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js)
        }
    }

    private func sendPurchaseCancelled() {
        let js = "window.onIosPurchaseCancelled && window.onIosPurchaseCancelled()"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js)
        }
    }

    private func sendPurchaseError(_ error: String) {
        let escapedError = error.replacingOccurrences(of: "'", with: "\\'")
        let js = "window.onIosPurchaseError && window.onIosPurchaseError('\(escapedError)')"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js)
        }
    }

    private func sendRestoreSuccess() {
        let js = "window.onIosRestoreSuccess && window.onIosRestoreSuccess()"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js)
        }
    }

    private func sendRestoreError(_ error: String) {
        let escapedError = error.replacingOccurrences(of: "'", with: "\\'")
        let js = "window.onIosRestoreError && window.onIosRestoreError('\(escapedError)')"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js)
        }
    }

    private func sendSubscriptionStatus(active: Bool) {
        let js = "window.onIosSubscriptionStatus && window.onIosSubscriptionStatus(\(active))"
        DispatchQueue.main.async {
            self.webView?.evaluateJavaScript(js)
        }
    }
}

File: android/app/src/main/java/com/myspeedpuzzling/billing/BillingBridge.kt
package com.myspeedpuzzling.billing

import android.content.Context
import android.webkit.JavascriptInterface
import android.webkit.WebView
import java.lang.ref.WeakReference

/**
 * JavaScript bridge for Google Play Billing on Android.
 * Receives purchase requests from web JavaScript and initiates native billing flows.
 */
class BillingBridge(context: Context) {
    private val contextRef = WeakReference(context)
    private var webViewRef: WeakReference<WebView>? = null
    private var billingManager: BillingManager? = null

    init {
        billingManager = BillingManager(context).apply {
            setCallback(object : BillingManager.BillingCallback {
                override fun onPurchaseSuccess(productId: String, purchaseToken: String) {
                    sendPurchaseSuccess(productId, purchaseToken)
                }

                override fun onPurchaseCancelled() {
                    sendPurchaseCancelled()
                }

                override fun onPurchaseError(error: String) {
                    sendPurchaseError(error)
                }

                override fun onRestoreSuccess() {
                    sendRestoreSuccess()
                }

                override fun onRestoreError(error: String) {
                    sendRestoreError(error)
                }

                override fun onSubscriptionStatus(active: Boolean) {
                    sendSubscriptionStatus(active)
                }
            })
        }
    }

    fun setWebView(webView: WebView) {
        this.webViewRef = WeakReference(webView)
    }

    @JavascriptInterface
    fun purchase(productId: String) {
        billingManager?.launchPurchaseFlow(productId)
    }

    @JavascriptInterface
    fun restorePurchases() {
        billingManager?.restorePurchases()
    }

    @JavascriptInterface
    fun checkSubscription() {
        billingManager?.checkSubscriptionStatus()
    }

    fun destroy() {
        billingManager?.destroy()
    }

    // JavaScript callbacks

    private fun sendPurchaseSuccess(productId: String, purchaseToken: String) {
        val webView = webViewRef?.get() ?: return
        val js = "window.onAndroidPurchaseSuccess && window.onAndroidPurchaseSuccess('$productId', '$purchaseToken')"
        webView.post {
            webView.evaluateJavascript(js, null)
        }
    }

    private fun sendPurchaseCancelled() {
        val webView = webViewRef?.get() ?: return
        val js = "window.onAndroidPurchaseCancelled && window.onAndroidPurchaseCancelled()"
        webView.post {
            webView.evaluateJavascript(js, null)
        }
    }

    private fun sendPurchaseError(error: String) {
        val webView = webViewRef?.get() ?: return
        val escapedError = error.replace("'", "\\'")
        val js = "window.onAndroidPurchaseError && window.onAndroidPurchaseError('$escapedError')"
        webView.post {
            webView.evaluateJavascript(js, null)
        }
    }

    private fun sendRestoreSuccess() {
        val webView = webViewRef?.get() ?: return
        val js = "window.onAndroidRestoreSuccess && window.onAndroidRestoreSuccess()"
        webView.post {
            webView.evaluateJavascript(js, null)
        }
    }

    private fun sendRestoreError(error: String) {
        val webView = webViewRef?.get() ?: return
        val escapedError = error.replace("'", "\\'")
        val js = "window.onAndroidRestoreError && window.onAndroidRestoreError('$escapedError')"
        webView.post {
            webView.evaluateJavascript(js, null)
        }
    }

    private fun sendSubscriptionStatus(active: Boolean) {
        val webView = webViewRef?.get() ?: return
        val js = "window.onAndroidSubscriptionStatus && window.onAndroidSubscriptionStatus($active)"
        webView.post {
            webView.evaluateJavascript(js, null)
        }
    }
}
